### PR TITLE
WIP: Include some tests of store equality

### PR DESCRIPTION
**Pull request by jakirkham**
Provide a few tests that check to see if different stores compare equal or not.

Note: Currently this fails for a few stores. Hopefully this will be a good point to debug and fix those.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [ ] Changes documented in docs/release.rst
* [ ] Docs build locally (e.g., run ``tox -e docs``)
* [ ] AppVeyor and Travis CI passes
* [ ] Test coverage is 100% (Coveralls passes)

### CODE REVIEW

**Comment by alimanfoo** [comment]
Thanks @jakirkham. FWIW we might need to think a bit about what it should mean for two stores to compare equal. Currently, for example, two ``DirectoryStore`` objects will [compare equal](https://github.com/zarr-developers/zarr/blob/master/zarr/storage.py#L781) if they both use the same directory path on the file system. Consequently, two ``DirectoryStore`` objects that use different directory paths will always compare unequal, regardless of what data are stored there. Also, a ``DirectoryStore`` object and a store object of a different class (e.g., ``ZipStore``) will always compare unequal, regardless of what data are stored. This seemed like a simple and sensible thing to do at the time I wrote those storage classes, but I'll need to go back and try to remember why. Looking now to find what tests cover the current store ``__eq__()`` implementations...

**Comment by alimanfoo** [comment]
So I'm pretty sure that the ``__eq__()`` implementations on the store classes are not currently tested directly, however they will be used inside the [``Array.__eq__()`` implementation](https://github.com/zarr-developers/zarr/blob/master/zarr/core.py#L416) and in the [``Group.__eq__()`` implementation](https://github.com/zarr-developers/zarr/blob/master/zarr/hierarchy.py#L184). Now trying to find where those are tested...

**Comment by alimanfoo** [comment]
OK, ``Group`` equality is used in several tests in the test_hierarchy module, e.g., [test_require_group](https://github.com/zarr-developers/zarr/blob/master/zarr/tests/test_hierarchy.py#L168), [test_getitem_contains_iterator](https://github.com/zarr-developers/zarr/blob/master/zarr/tests/test_hierarchy.py#L406), and [test_paths](https://github.com/zarr-developers/zarr/blob/master/zarr/tests/test_hierarchy.py#L791). 

Struggling to find where ``Array.__eq__()`` is tested...

**Comment by alimanfoo** [comment]
OK, got it. We currently rely on ``Array.__eq__()`` in several tests within the test_hierarchy module, e.g., [here](https://github.com/zarr-developers/zarr/blob/master/zarr/tests/test_hierarchy.py#L648), [here](https://github.com/zarr-developers/zarr/blob/master/zarr/tests/test_hierarchy.py#L423), [here](https://github.com/zarr-developers/zarr/blob/master/zarr/tests/test_hierarchy.py#L308). We don't currently test it within the test_core module.

**Comment by alimanfoo** [comment]
Reviewing the current tests, I think the current implementations of ``__eq__()`` in the ``Array``, ``Group`` and store classes do serve the tests well. 

Is there a reason for considering a different behaviour for the store ``__eq__()`` implementations?

**Comment by jakirkham** [comment]
Based on our discussions in issue ( https://github.com/zarr-developers/zarr/issues/349 ) and the problem seen in issue ( https://github.com/zarr-developers/zarr/issues/348 ), it does seem like we need to start testing whether stores compare equal and if this is as expected to catch and remove bugs.

**Comment by alimanfoo** [comment]
> Based on our discussions in issue ( #349 ) and the problem seen in issue ( #348 ), it does seem like we need to start testing whether stores compare equal and if this is as expected to catch and remove bugs.

FWIW if we switch to using DictStore as the default store, then #348 should go away. I.e., we will get consistent behaviour for array equals comparison across all stores, which is essentially that two arrays will compare equal iff they share the same storage.

**Comment by jakirkham** [comment]
So I've made some changes inline with the intent. There will still be a failure for the `dict` case that we can discuss.

**Comment by jakirkham** [comment]
> So I'm pretty sure that the `__eq__()` implementations on the store classes are not currently tested directly, however they will be used inside the [`Array.__eq__()` implementation](https://github.com/zarr-developers/zarr/blob/master/zarr/core.py#L416) and in the [`Group.__eq__()` implementation](https://github.com/zarr-developers/zarr/blob/master/zarr/hierarchy.py#L184).

I'm not quite following this. Wouldn't we want to ensure that two Arrays have the same content regardless of where they are stored? Would also be curious why wouldn't we want to compare Groups' contents as well. Thoughts?

**Comment by alimanfoo** [comment]
On Wed, 13 Feb 2019 at 08:47, jakirkham <notifications@github.com> wrote:

> So I'm pretty sure that the __eq__() implementations on the store classes
> are not currently tested directly, however they will be used inside the
> Array.__eq__() implementation
> <https://github.com/zarr-developers/zarr/blob/master/zarr/core.py#L416>
> and in the Group.__eq__() implementation
> <https://github.com/zarr-developers/zarr/blob/master/zarr/hierarchy.py#L184>
> .
>
> I'm not quite following this. Wouldn't we want to ensure that two Arrays
> have the same content regardless of where they are stored?
>
The tests I linked to currently use __eq__ to verify that two arrays draw
their data from the same store (or two groups draw their data from the same
store). They are not trying to compare the actual data. Part of that
behaviour involves testing that two store objects are "equal" in the sense
that they both draw their data from the same source (e.g., the same
directory on a file system). Again that doesn't involve comparing any
actual data. Those tests all need some way of making this kind of
comparison, although perhaps __eq__ is not the right method for it.

I could imagine in other situations you might want to compare the actual
data contents of two arrays (or two groups or two stores) and return true
if the data are the same, regardless if drawn from the same data source or
not. That's a separate requirement. Maybe __eq__ should behave like that.
In this case, should zarr arrays behave like numpy arrays, and return an
array of boolean values with broadcasting etc.?

Bottom line, I think there are two separate requirements for arrays: (1)
compare two arrays and return True if they are "the same" in the sense that
they both draw their data from the same source; and (2) compare the data
contents of two arrays and return something.

Similarly, I think there are two separate requirements for groups: (1)
return True if two groups draw their data from the same source; (2) compare
the contents of two groups.

Similarly for stores: (1) return True if two store objects draw their data
from the same underlying source; (2) compare the contents of two stores.

Now the question is, what should __eq__ do? And what other methods do we
need to cover any remaining requirements not covered by __eq__?
